### PR TITLE
Replace AST cache with in-memory program storage, update to Cadence v0.13.0

### PR DESCRIPTION
**Pull request by turbolent**
- Update to Cadence v0.13.0
- Update to Go SDK v0.15.0
- Remove the AST cache
- Implement a thread-safe, in-memory program store. This cannot be simply a typical cache, as during execution of transactions/scripts, Cadence expects results to be consistently returned, i.e. if it called `SetProgram`, it expects `GetProgram` to return what it set, which isn't ensured by a plain cache
- Parallelize some unit tests

### CODE REVIEW

**Review by ramtinms — CHANGES_REQUESTED** [review]
This breaks the verification nodes, the cache needs to touch the registers at the location through the view.

**Comment by turbolent** [comment]
@ramtinms can you elaborate what that means and what needs to change where?

**Review by m4ksio — APPROVED** [review]
Nice and clean!

**Comment by turbolent** [comment]
Isn't this what this part ensures? https://github.com/onflow/flow-go/pull/339/files#diff-b4e3be012ea5724c7d1faf9e1ffa2780351d5c876037ccd25ee33704d587e676R276-R280

**Review by ramtinms — APPROVED** [review]
Right I missed that part somehow, I always assumed cache handles that logic but forgot about the env add extra functionality.

**Review by SupunS — APPROVED** [review]

**Comment by turbolent** [comment]
This still needs two changes:

- [ ] During execution of a transaction or script, Cadence will store (`SetProgram`) the program for the transaction/script and also read it (`GetProgram`). After the execution this should be cleaned up, we should not persist these programs across multiple transaction / script executions, only programs with address locations (contracts)
- [ ] Changes to the program storage should have no affect when a script is executed or a transaction is reverted. Maybe we can have a two-tiered system here where changes of an execution are only written to a temporary storage and those changes are only applied to the long-term storage for successful transactions

@ramtinms thoughts?

File: fvm/fvm_test.go [hunk 1]
```diff
@@ -207,16 +207,18 @@ func TestBlockContext_ExecuteTransaction(t *testing.T) {
...
 
 	vm := fvm.New(rt)
 
-	cache, err := fvm.NewLRUASTCache(CacheSize)
-	require.NoError(t, err)
-
-	ctx := fvm.NewContext(zerolog.Nop(), fvm.WithChain(chain), fvm.WithASTCache(cache), fvm.WithCadenceLogging(true))
+	ctx := fvm.NewContext(
+		zerolog.Nop(),
+		fvm.WithChain(chain), fvm.WithCadenceLogging(true))
```
Comment (turbolent) [inline fvm/fvm_test.go:221]:
```suggestion
		zerolog.Nop(),
		fvm.WithChain(chain), 
		fvm.WithCadenceLogging(true),
	)
```

File: fvm/transactionFeeDeductor.go [hunk 2]
```diff
@@ -18,17 +18,20 @@ func (d *TransactionFeeDeductor) Process(
...
-	if err == nil {
-		er := st.Commit()
-		if er != nil {
-			panic(er)
-		}
-		er = ctx.Programs.Commit()
-		if er != nil {
-			panic(er)
-		}
+	if err != nil {
```
Comment (turbolent) [inline fvm/transactionFeeDeductor.go:21]: Early exit on error instead of nesting

File: engine/execution/computation/computer/computer_test.go [hunk 3]
```diff
@@ -229,32 +240,139 @@ func TestBlockExecutor_ExecuteBlock(t *testing.T) {
...
+
+		require.Equal(t, 0, execCtx.Programs.Length())
+	})
 }
 
-type eventEmittingRuntime struct {
-	events [][]cadence.Event
+type testRuntime struct {
+	executeScript      func(runtime.Script, runtime.Context) (cadence.Value, error)
+	executeTransaction func(runtime.Script, runtime.Context) error
```
Comment (turbolent) [inline engine/execution/computation/computer/computer_test.go:360]: Generalize into a value that can be used in multiple different tests

File: engine/execution/computation/computer/computer.go [hunk 4]
```diff
@@ -47,14 +45,9 @@ func NewBlockComputer(
...
 	logger zerolog.Logger,
 ) (BlockComputer, error) {
-	systemChunkASTCache, err := fvm.NewLRUASTCache(SystemChunkASTCacheSize)
-	if err != nil {
-		return nil, fmt.Errorf("cannot create system chunk AST cache: %w", err)
-	}
 
 	systemChunkCtx := fvm.NewContextFromParent(
 		vmCtx,
-		fvm.WithASTCache(systemChunkASTCache),
```
Comment (ramtinms) [inline engine/execution/computation/computer/computer.go:57]: 👍

File: fvm/programs.go [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+}
+
+func NewPrograms(parent *Programs) *Programs {
+	return &Programs{
+		parent:   parent,
+		programs: map[common.LocationID]ProgramEntry{},
+	}
+}
+
+func (p *Programs) Get(locationID common.LocationID) *interpreter.Program {
```
Comment (ramtinms) [inline fvm/programs.go:28]: I think might be better to change this to Get(location common.Location) and we do the cast to ID here.

File: fvm/programs.go [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+		// and not programs for transactions or scripts
+
+		if _, ok := programEntry.Location.(common.AddressLocation); !ok {
+			continue
+		}
+
+		parent.Set(
+			programEntry.Location,
+			programEntry.Program,
+		)
```
Comment (ramtinms) [inline fvm/programs.go:83]: Maybe adding a note here that this is not update safe right now due to dependencies

File: fvm/programs.go [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+	if p.parent != nil {
+		return p.parent.Get(locationID)
+	}
+
+	// The program is neither available here nor in the parent
+
+	return nil
+}
+
+func (p *Programs) Set(location common.Location, program *interpreter.Program) {
```
Comment (ramtinms) [inline fvm/programs.go:48]: Any value to add the logic here to when we set an already existing value we purge the map ?

**Review by ramtinms — APPROVED** [review]
Looks good to me, just few minor comments. Also I think it might be valuable to add an extra flag to be able to disable the cache, in case we had to.

**Review by psiemens — APPROVED** [review]
This is great, nice work! 👏 

The logic itself looks sound.

I'm good with merging this PR as-is since I know it's time sensitive, but I also wanted point out a small design point. 

`fvm.Context` was originally designed to be a stateless object that could be used to configure the runtime environment. The AST cache already broke that statelessness, so this PR doesn't introduce any significant changes, but I do worry about the fact that an `fvm.Context` instance can now exhibit different behaviour depending on the internal state of  the `Programs` field (e.g. depending on whether `Programs.Commit()` is called).

An easy fix would be to move the lifecycle management of `fvm.Programs` out of `fmv.Context`, e.g.:

```go
parentPrograms := fvm.NewPrograms()

parentContext := fvm.NewContext(fvm.WithPrograms(programs))

// ...

childPrograms := fvm.NewProgramsFromParent(parentPrograms)

childCtx := fvm.NewContext(fvm.WithPrograms(childPrograms))

// ...

parentPrograms.Commit(childPrograms)
```

However, I'm now starting to wonder if it makes sense to turn `fvm.Context` into a stateful object that manages the branch/commit logic of both `Programs` and the register state views (`state.View`), too. It might be simpler to have a single object that encapsulates all of the execution logic for a single branch. I'll think about this more, but just wanted to bring it up, no need to change anything about this PR.

File: go.mod [hunk 8]
```diff
@@ -33,7 +33,7 @@ require (
 	github.com/libp2p/go-tcp-transport v0.2.1
 	github.com/m4ksio/wal v1.0.0
 	github.com/multiformats/go-multiaddr v0.3.1
-	github.com/onflow/cadence v0.12.5
+	github.com/onflow/cadence v0.12.6-0.20210129000118-ad2b07a09690
```
Comment (janezpodhostnik) [inline go.mod:36]: Replace with tag if possible

**Review by janezpodhostnik — APPROVED** [review]
Nice tests!

**Comment by turbolent** [comment]
Given that we might have a bit more time to implement this properly: 
What would it take to make this properly fork-aware? Any hints on where to start?

cc @m4ksio @ramtinms

File: fvm/programs.go (hunk 5 above)
Comment (turbolent) [reply fvm/programs.go]: Good idea, changed in 694e65499b1ac4447d4623a892353652aa5a66b8

**Comment by turbolent** [comment]
I'm going to simplify this now so we restore the current behaviour: programs are only cached during the execution of a single transaction or script, not yet across collections and blocks.

**Comment by turbolent** [comment]
Switched to a feature branch for Cadence v0.13. I'll release that today and then update the dependency here (instead of depending on a commit)

File: engine/execution/computation/manager_test.go [hunk 9]
```diff
@@ -103,3 +104,40 @@ func TestComputeBlockWithStorage(t *testing.T) {
 	require.Len(t, returnedComputationResult.StateSnapshots, 1+1) // 1 coll + 1 system chunk
 	assert.NotEmpty(t, returnedComputationResult.StateSnapshots[0].Delta)
 }
+
+func TestExecuteScript(t *testing.T) {
+
+	t.Run("scripts do not store programs", func(t *testing.T) {
```
Comment (janezpodhostnik) [inline engine/execution/computation/manager_test.go:110]: I can't tell which line of this tests asserts that "scripts do not store programs"...

**Review by janezpodhostnik — APPROVED** [review]
LGTM. Just had one small question about a test.

File: engine/execution/computation/manager_test.go (hunk 9 above)
Comment (turbolent) [reply engine/execution/computation/manager_test.go]: Ah, that was a left-over from earlier code. The test ensured that programs that were stored during script execution would not be committed, i.e. used for other scripts or even transactions.

The PR now does not implement this anymore, so I removed the confusing subtest name. The test of executing a script on its own is still IMHO useful, so I kept it.

**Review by ramtinms — APPROVED** [review]
this looks good to me, we just need to implement the fork-aware cache in the next PR and update the fvm/programs.go to use the one passed by the execution env.

**Comment by turbolent** [comment]
@ramtinms Right, I simplified this PR so we can the update to the latest version of Cadence without having to wait on the fork-aware cross-TX caching, that will be implemented in dapperlabs/flow-go#5282

**Pull request merged**